File: Sources/BNKKit/Logs.swift
import Foundation

/// One line out of one container.
public struct LogLine: Identifiable, Sendable, Hashable {
    public let id = UUID()
    public let at: Date?
    public let pod: String
    public let container: String?
    public let text: String
    public let level: Level

    public enum Level: String, Sendable, CaseIterable {
        case error, warning, info

        /// Guessed from the shapes that actually turn up in these clusters,
        /// which is more than one convention: klog's `E0831 09:00:43`, logfmt's
        /// `level=warn`, JSON's `"level":"error"`, and F5's CWC logger, which
        /// spells it `"l"="error"` — a single letter, and the reason the first
        /// version of this rendered a rabbitmq connection failure and a Redis
        /// retry storm in the same colour as a routine status line.
        ///
        /// It is a guess and will sometimes be wrong. The bias is deliberate:
        /// a missed error is invisible, while a line wrongly painted amber is
        /// still readable.
        public static func guessed(from line: String) -> Level {
            if let klog = klogLevel(line) { return klog }
            let head = line.prefix(400).lowercased()
            for marker in ["\"l\"=\"error\"", "\"l\"=\"critical\"", "\"l\"=\"fatal\"",
                           "level=error", "level=fatal", "level=critical",
                           "\"level\":\"error\"", "\"level\":\"fatal\"", "[error]", "[fatal]"]
            where head.contains(marker) { return .error }
            for marker in ["\"l\"=\"warn\"", "level=warn", "\"level\":\"warn\"", "[warn]"]
            where head.contains(marker) { return .warning }
            if contains(word: "panic", in: head) || contains(word: "fatal", in: head) { return .error }
            if contains(word: "error", in: head) || contains(word: "errors", in: head) { return .error }
            for word in ["failed", "failure", "refused", "unreachable", "timeout", "warning"]
            where contains(word: word, in: head) { return .warning }
            return .info
        }

        /// `E0831 09:00:43.931896` — severity letter, then a four-digit date.
        private static func klogLevel(_ line: String) -> Level? {
            guard line.count > 5, let first = line.first,
                  line.dropFirst().prefix(4).allSatisfy(\.isNumber) else { return nil }
            switch first {
            case "E", "F": return .error
            case "W":      return .warning
            case "I":      return .info
            default:       return nil
            }
        }

        /// Whole word only, so `errorless` does not count — and neither does a
        /// path component, which is why `/`, `-` and `_` are treated as part of
        /// a word rather than as boundaries: `/var/log/failed/` is a directory,
        /// not a fault. Nor does "0 errors", the reassuring case that a naive
        /// substring search turns into an alarm.
        private static func contains(word: String, in haystack: some StringProtocol) -> Bool {
            var index = haystack.startIndex
            while let range = haystack.range(of: word, range: index..<haystack.endIndex) {
                let beforeIndex = range.lowerBound
                let before: Character? = beforeIndex == haystack.startIndex
                    ? nil : haystack[haystack.index(before: beforeIndex)]
                let after: Character? = range.upperBound == haystack.endIndex
                    ? nil : haystack[range.upperBound]
                func isPartOfAWord(_ c: Character) -> Bool {
                    c.isLetter || c.isNumber || c == "/" || c == "-" || c == "_"
                }
                let boundedBefore = before.map { !isPartOfAWord($0) } ?? true
                let boundedAfter = after.map { !isPartOfAWord($0) } ?? true
                if boundedBefore, boundedAfter, !negated(before: beforeIndex, in: haystack) { return true }
                index = range.upperBound
            }
            return false
        }

        /// "no errors" and "0 errors" are the opposite of an error.
        private static func negated(before index: some Comparable, in haystack: some StringProtocol) -> Bool {
            guard let i = index as? String.Index else { return false }
            let prefix = haystack[haystack.startIndex..<i].suffix(6).lowercased()
            return prefix.hasSuffix("no ") || prefix.hasSuffix("0 ") || prefix.hasSuffix("zero ")
        }
    }
}

extension KubeClient {
    /// Follow one container's log.
    ///
    /// Nothing is installed to make this work: a container's stdout is captured
    /// by the runtime on the node, and the apiserver will stream it back through
    /// the kubelet. That is the same door everything else here goes through, and
    /// it is why logs need no collector in the cluster — only history does.
    ///
    /// `timestamps=true` because merging several pods into one view needs
    /// something to sort by, and the arrival order of concurrent streams is not
    /// it.
    public func logStream(namespace: String, pod: String, container: String? = nil,
                          tailLines: Int = 80, follow: Bool = true)
    -> AsyncThrowingStream<LogLine, Error> {
        AsyncThrowingStream { continuation in
            let work = Task {
                do {
                    var query = [
                        "follow": follow ? "true" : "false",
                        "tailLines": String(tailLines),
                        "timestamps": "true",
                    ]
                    if let container { query["container"] = container }
                    let request = try self.request(
                        "GET", "/api/v1/namespaces/\(namespace)/pods/\(pod)/log", query: query)
                    let (bytes, response) = try await self.streamingSession.bytes(for: request)
                    let code = (response as? HTTPURLResponse)?.statusCode ?? 0
                    guard (200..<300).contains(code) else {
                        throw Failure.http(code, "could not read logs for \(pod)")
                    }
                    for try await raw in bytes.lines {
                        if Task.isCancelled { break }
                        continuation.yield(Self.parse(raw, pod: pod, container: container))
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in work.cancel() }
        }
    }

    /// `2026-08-31T09:24:18.442Z the rest of the line`
    static func parse(_ raw: String, pod: String, container: String?) -> LogLine {
        var at: Date?
        var text = raw
        if let space = raw.firstIndex(of: " ") {
            let stamp = String(raw[raw.startIndex..<space])
            if let date = rfc3339.date(from: stamp) ?? rfc3339Fractional.date(from: stamp) {
                at = date
                text = String(raw[raw.index(after: space)...])
            }
        }
        return LogLine(at: at, pod: pod, container: container,
                       text: text, level: .guessed(from: text))
    }

    /// `nonisolated(unsafe)` because `ISO8601DateFormatter` is not `Sendable`,
    /// although Foundation's date formatters have been documented as safe to use
    /// from several threads for a long time. Both are configured once here and
    /// never mutated afterwards, only asked to parse.
    nonisolated(unsafe) static let rfc3339: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    nonisolated(unsafe) static let rfc3339Fractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
}

File: dart/bnk_kit/lib/src/logs.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'kube_client.dart';
import 'time.dart';

enum LogLevel {
  error,
  warning,
  info;

  /// Guessed from the shapes that actually turn up in these clusters, which
  /// is more than one convention: klog's `E0831 09:00:43`, logfmt's
  /// `level=warn`, JSON's `"level":"error"`, and F5's CWC logger, which spells
  /// it `"l"="error"`, a single letter, and the reason the first version of
  /// this rendered a rabbitmq connection failure and a Redis retry storm in
  /// the same colour as a routine status line.
  ///
  /// It is a guess and will sometimes be wrong. The bias is deliberate: a
  /// missed error is invisible, while a line wrongly painted amber is still
  /// readable.
  static LogLevel guessed(String line) {
    final klog = _klogLevel(line);
    if (klog != null) return klog;
    final head =
        (line.length > 400 ? line.substring(0, 400) : line).toLowerCase();
    for (final marker in const [
      '"l"="error"', '"l"="critical"', '"l"="fatal"',
      'level=error', 'level=fatal', 'level=critical',
      '"level":"error"', '"level":"fatal"', '[error]', '[fatal]',
    ]) {
      if (head.contains(marker)) return error;
    }
    for (final marker in const [
      '"l"="warn"', 'level=warn', '"level":"warn"', '[warn]',
    ]) {
      if (head.contains(marker)) return warning;
    }
    if (_containsWord('panic', head) || _containsWord('fatal', head)) {
      return error;
    }
    if (_containsWord('error', head) || _containsWord('errors', head)) {
      return error;
    }
    for (final word in const [
      'failed', 'failure', 'refused', 'unreachable', 'timeout', 'warning',
    ]) {
      if (_containsWord(word, head)) return warning;
    }
    return info;
  }

  static final _fourDigits = RegExp(r'^\d{4}$');

  /// `E0831 09:00:43.931896`: severity letter, then a four-digit date.
  static LogLevel? _klogLevel(String line) {
    if (line.length <= 5) return null;
    if (!_fourDigits.hasMatch(line.substring(1, 5))) return null;
    return switch (line[0]) {
      'E' || 'F' => error,
      'W' => warning,
      'I' => info,
      _ => null,
    };
  }

  static final _wordCharacter = RegExp(r'^[\p{L}\p{N}/_-]$', unicode: true);

  /// Whole word only, so `errorless` does not count, and neither does a path
  /// component, which is why `/`, `-` and `_` are treated as part of a word
  /// rather than as boundaries: `/var/log/failed/` is a directory, not a
  /// fault. Nor does "0 errors", the reassuring case that a naive substring
  /// search turns into an alarm.
  static bool _containsWord(String word, String haystack) {
    var index = 0;
    while (true) {
      final at = haystack.indexOf(word, index);
      if (at < 0) return false;
      final end = at + word.length;
      final boundedBefore =
          at == 0 || !_wordCharacter.hasMatch(haystack[at - 1]);
      final boundedAfter =
          end == haystack.length || !_wordCharacter.hasMatch(haystack[end]);
      if (boundedBefore && boundedAfter && !_negated(at, haystack)) return true;
      index = end;
    }
  }

  /// "no errors" and "0 errors" are the opposite of an error.
  static bool _negated(int index, String haystack) {
    final prefix =
        haystack.substring(index >= 6 ? index - 6 : 0, index).toLowerCase();
    return prefix.endsWith('no ') ||
        prefix.endsWith('0 ') ||
        prefix.endsWith('zero ');
  }
}

/// One line out of one container.
class LogLine {
  static int _next = 0;

  /// Stable identity for a list, which arrival order does not give.
  final int id = _next++;
  final DateTime? at;
  final String pod;
  final String? container;
  final String text;
  final LogLevel level;

  LogLine(
      {required this.at,
      required this.pod,
      required this.container,
      required this.text,
      required this.level});

  /// `2026-08-31T09:24:18.442Z the rest of the line`
  static LogLine parse(String raw, {required String pod, String? container}) {
    DateTime? at;
    var text = raw;
    final space = raw.indexOf(' ');
    if (space > 0) {
      final date = Rfc3339.parse(raw.substring(0, space));
      if (date != null) {
        at = date;
        text = raw.substring(space + 1);
      }
    }
    return LogLine(
        at: at,
        pod: pod,
        container: container,
        text: text,
        level: LogLevel.guessed(text));
  }
}

extension LogsApi on KubeClient {
  /// Follow one container's log.
  ///
  /// Nothing is installed to make this work: a container's stdout is captured
  /// by the runtime on the node, and the apiserver will stream it back through
  /// the kubelet. That is the same door everything else here goes through, and
  /// it is why logs need no collector in the cluster; only history does.
  ///
  /// `timestamps=true` because merging several pods into one view needs
  /// something to sort by, and the arrival order of concurrent streams is not
  /// it. Cancelling the subscription drops the connection.
  Stream<LogLine> logStream(
      {required String namespace,
      required String pod,
      String? container,
      int tailLines = 80,
      bool follow = true}) {
    final controller = StreamController<LogLine>();
    HttpClientRequest? request;
    StreamSubscription<String>? lines;
    var cancelled = false;

    controller.onListen = () async {
      try {
        final query = <String, Object>{
          'follow': follow ? 'true' : 'false',
          'tailLines': '$tailLines',
          'timestamps': 'true',
          if (container != null) 'container': container,
        };
        final req = await open('GET', '/api/v1/namespaces/$namespace/pods/$pod/log', query);
        request = req;
        final response = await req.close();
        if (response.statusCode < 200 || response.statusCode >= 300) {
          await response.drain<void>();
          throw KubeFailure.http(response.statusCode, 'could not read logs for $pod');
        }
        if (cancelled) {
          response.detachSocket().then((s) => s.destroy()).ignore();
          return;
        }
        lines = response
            .transform(utf8.decoder)
            .transform(const LineSplitter())
            .listen(
              (raw) => controller.add(LogLine.parse(raw, pod: pod, container: container)),
              onError: (Object e, StackTrace st) => controller.addError(e, st),
              onDone: () => controller.close(),
              cancelOnError: true,
            );
      } catch (e, st) {
        if (!controller.isClosed) {
          controller.addError(e, st);
          await controller.close();
        }
      }
    };
    controller.onCancel = () async {
      cancelled = true;
      final sub = lines;
      if (sub != null) {
        await sub.cancel();
      } else {
        request?.abort();
      }
    };
    return controller.stream;
  }
}
